src/Builder/BuilderEncoder.php

```diff
@@ -0,0 +1,181 @@
+            foreach ($value->fields ?? [] as $key => $val) {
+                $result->{$key} = $this->encodeIfSupported($val);
+            }
+
+            return $this->wrap($value, $result);
+        }
+
+        if ($value instanceof ProjectStage) {
+            $result = new stdClass();
+            // Specific: fields are encoded as a map of properties to their values at the top level as _id
+            foreach ($value->specifications as $key => $val) {
+                $result->{$key} = $this->encodeIfSupported($val);
+            }
+
+            return $this->wrap($value, $result);
+        }
+
+        if ($value instanceof OrQuery) {
+            $result = [];
+            foreach ($value->query as $query) {
+                $encodedQuery = new stdClass();
+                foreach ($query as $field => $expression) {
+                    // Specific: $or queries are encoded as a list of expressions
+                    // We need to merge query expressions into a single object
+                    if (is_array($expression) && array_is_list($expression)) {
+                        $mergedExpressions = [];
+                        foreach ($expression as $expr) {
+                            $mergedExpressions = array_merge($mergedExpressions, (array) $this->encodeIfSupported($expr));
+                        }
+
+                        $encodedQuery->{$field} = (object) $mergedExpressions;
+                    } else {
+                        $encodedQuery->{$field} = $this->encodeIfSupported($expression);
+                    }
+                }
+
+                $result[] = $encodedQuery;
+            }
+
+            return $this->wrap($value, $result);
+        }
+
+        // The generic but incomplete encoding code
+        switch ($value::ENCODE) {
+            case self::ENCODE_AS_SINGLE:
+                return $this->encodeAsSingle($value);
+
+            case self::ENCODE_AS_ARRAY:
+                return $this->encodeAsArray($value);
+
+            case self::ENCODE_AS_OBJECT:
+                return $this->encodeAsObject($value);
+        }
+
+        throw new LogicException(sprintf('Class "%s" does not have a valid ENCODE constant.', $value::class));
+    }
+
+    private function encodeAsArray(ExpressionInterface|StageInterface $value): stdClass
+    {
+        $result = [];
+        /** @var mixed $val */
+        foreach (get_object_vars($value) as $val) {
+            $result[] = $this->encodeIfSupported($val);
+        }
+
+        return $this->wrap($value, $result);
+    }
+
+    private function encodeAsObject(ExpressionInterface|StageInterface $value): stdClass
+    {
+        $result = new stdClass();
+        /** @var mixed $val */
+        foreach (get_object_vars($value) as $key => $val) {
+            /** @var mixed $val */
+            $val = $this->encodeIfSupported($val);
+            // @todo check for undefined value vs null
```

Comment:
Would "undefined" be considered a sentinel value? If so, I was recently talking about this in the phpc (not ours) Discord with @crell and he suggested using a single-value enum.

![image](https://github.com/mongodb/mongo-php-library/assets/244663/86cdad75-f492-494e-92a3-bb4718c323ea)

This lead me to an interesting wiki article: [Elephant in Cairo](https://en.wikipedia.org/wiki/Elephant_in_Cairo)

Reply:
I added the `Optional` enum, with `Undefined` case. I like that "optional" is self-documenting that a parameter is optional.